### PR TITLE
PIM-6773: highlight incomplete values in the product model edit form

### PR DESCRIPTION
**Pull request by jjanvier**
This PR allows to highlight missing values in the product model edit form. Just like what we have in the product edit form.

In the product edit form, this is done by normalizing the _completenesses_ of the product.
But the product model has no completenesses. To sum functionally:
- product => completeness = missing values & ratios
- product model => CONCEPT WITH NO NAME = missing values & number of complete products / number of products

So we wanted to extract the most possible logic from the `Pim\Component\Catalog\Completeness\CompletenessCalculator` to be able to use it for the product models. We extracted:
- the computation of required values for an entity with values (ie: product or product model)
- the computation of incomplete values for an entity with values depending on its required values

In the backend, it's OK. The problem reminds in the frontend. As the product model edit form uses the same logic than the product form, we have to normalize those incomplete values information the same.
Which is why the `ProductModelIncompleteValuesNormalizer` is the a kind of mix and copy paste of both `CompletenessCalculator` and `CompletenessCollectionNormalizer`. This is not very elegant, but it works and allows to isolate the :hankey: without the UI normalization without introducing a non existent "completeness" concept on the product models.

TODO in this PR
- [x] move the required values stuff in a proper directory
- [x] move the incomplete values stuff in a proper directory
- [ ] changelog
- [ ] maybe rename `CompletenessCalculator` as now it's simply a `CompletenessCollectionFactory`
- [x] spec `ProductModelIncompleteValuesNormalizer` 

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | Y
| Added integration tests           | Y
| Changelog updated                 | TODO
| Review and 2 GTM                  | -
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Component/Catalog/Model/ValueCollectionInterface.php [hunk 1]
```diff
@@ -91,6 +91,16 @@ public function removeByAttribute(AttributeInterface $attribute);
...
 
     /**
+     * Get the value with the same attribute, channel and locale than $value.
+     * Or null if such a value does not exist.
+     *
+     * @param ValueInterface $value
+     *
+     * @return ValueInterface|null
+     */
+    public function getSame(ValueInterface $value);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Model/ValueCollectionInterface.php]: to remove if unused

File: src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+                    $attribute = $attributeRequirement->getAttribute();
+                    if ($attribute->isLocaleSpecific() && !$attribute->hasLocaleSpecific($locale)) {
+                        continue;
+                    }
+
+                    $channelCode = $attribute->isScopable() ? $channel->getCode() : null;
+                    $localeCode = $attribute->isLocalizable() ? $locale->getCode() : null;
+
+                    $requiredValues->add(new RequiredValue($attribute, $channelCode, $localeCode));
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php]: 👍

File: features/product-model/edit_product_model.feature [hunk 3]
```diff
@@ -77,3 +77,13 @@ Feature: Edit a product model
     When I press the "Save" button
     Then I should see validation tooltip "This value should be a valid number."
     And there should be 1 error in the "ERP" tab
+
+  Scenario: Quickly view missing values to fill in
+    Given I am logged in as "Mary"
+    And I am on the "amor" product model page
```
Comment (SamirBoulil) [inline features/product-model/edit_product_model.feature]: When

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            $requiredValuesForChannel = $this->requiredValueCollectionFactory->forChannel($family, $channel);
+
+            $kindOfCompleteness = [
+                'channel' => $channel->getCode(),
+                'labels' => $this->getChannelLabels($channel),
+                'locales' => []
+            ];
+
+            foreach ($channel->getLocales() as $locale) {
+                $requiredValues = $requiredValuesForChannel->filterByChannelAndLocale($channel, $locale);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php]: You shouldn't need to pass the channel here, as the requiredValueCollectionFactory->forChannel() calculates the requiredvalues for one channel. Maybe the channel should be part of the RequiredValueCollection ?

File: src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php [hunk 5]
```diff
@@ -18,73 +18,67 @@
...
- * according to the product family requirements. Then, it compares this
- * collection of fake values with the real values of the product, and generates
- * a list of completenesses, one completeness for each channel/locale possible
- * combinations.
- *
  * @author    Damien Carcel (damien.carcel@akeneo.com)
  * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
  */
 class CompletenessCalculator implements CompletenessCalculatorInterface
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php]: Munch cleanup 💪

File: src/Pim/Component/Catalog/Completeness/RequiredValue.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+use Pim\Component\Catalog\Model\AbstractValue;
+use Pim\Component\Catalog\Model\AttributeInterface;
+use Pim\Component\Catalog\Model\ValueInterface;
+
+/**
+ * A required value is the translation of the attribute requirements
+ * {@see Pim\Component\Catalog\Model\AttributeRequirementInterface} in terms of values.
+ *
+ * Therefore a required value contains no data. It simply express the fact that for instance:
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Completeness/RequiredValue.php]: expresses

File: src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @return RequiredValueCollectionInterface
+     */
+    public function forChannel(FamilyInterface $family, ChannelInterface $channel): RequiredValueCollectionInterface
+    {
+        $requiredValues = new RequiredValueCollection([]);
+
+        foreach ($this->filterRequirementsByChannel($family, $channel) as $attributeRequirement) {
+            foreach ($attributeRequirement->getChannel()->getLocales() as $locale) {
+                if ($attributeRequirement->isRequired()) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php]: could be done 1 line above

File: src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php (hunk 2 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/RequiredValueCollectionFactory.php]: add method is not necessary

File: src/Pim/Component/Catalog/Completeness/RequiredValueCollectionInterface.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * Returns all the elements of this collection that satisfy the given channel and locale.
+     *
+     * @param ChannelInterface $channel
+     * @param LocaleInterface  $locale
+     *
+     * @return RequiredValueCollectionInterface
+     */
+    public function filterByChannelAndLocale(
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/RequiredValueCollectionInterface.php]: filter?

File: src/Pim/Component/Catalog/Completeness/RequiredValueCollectionInterface.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * A collection of required values depending on the attribute requirements of a family.
+ * {@see Pim\Component\Catalog\Completeness\RequiredValue}
+ *
+ * @author    Julien Janvier <j.janvier@gmail.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+interface RequiredValueCollectionInterface extends \Countable, \IteratorAggregate
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/RequiredValueCollectionInterface.php]: interface or not?

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollection.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    public function hasSame(ValueInterface $value): bool
+    {
+        return array_key_exists($this->buildInternalKey($value), $this->values);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getAttributes(): Collection
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollection.php]: attributes

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollection.php [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * @param ValueInterface $value
+     *
+     * @return string
+     */
+    private function buildInternalKey(ValueInterface $value): string
+    {
+        $attribute = $value->getAttribute();
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollection.php]: mutualise in an object?

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * @param ValueCompleteCheckerInterface $completeValueChecker
+     */
+    public function __construct(ValueCompleteCheckerInterface $completeValueChecker)
+    {
+        $this->completeValueChecker = $completeValueChecker;
+    }
+
+    /**
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php]: doc

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        LocaleInterface $locale,
+        EntityWithValuesInterface $entityWithValues
+    ): IncompleteValueCollectionInterface {
+
+        $requiredValuesForChannelAndLocale = $requiredValues->filterByChannelAndLocale($channel, $locale);
+        $incompleteValues = new IncompleteValueCollection([]);
+
+        foreach ($requiredValuesForChannelAndLocale as $requiredValue) {
+            if ($this->isValueMissingOrIncomplete($requiredValue, $channel, $locale, $entityWithValues)) {
+                $incompleteValues->add($requiredValue);
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php]: drop add method

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        ChannelInterface $channel,
+        LocaleInterface $locale,
+        EntityWithValuesInterface $entityWithValues
+    ): IncompleteValueCollectionInterface {
+
+        $requiredValuesForChannelAndLocale = $requiredValues->filterByChannelAndLocale($channel, $locale);
+        $incompleteValues = new IncompleteValueCollection([]);
+
+        foreach ($requiredValuesForChannelAndLocale as $requiredValue) {
+            if ($this->isValueMissingOrIncomplete($requiredValue, $channel, $locale, $entityWithValues)) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php]: isMissingOrEmpty

File: src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+use Akeneo\Component\StorageUtils\Repository\IdentifiableObjectRepositoryInterface;
+use Pim\Component\Catalog\Completeness\Checker\ValueCompleteCheckerInterface;
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Pim\Component\Catalog\Model\EntityWithValuesInterface;
+use Pim\Component\Catalog\Model\LocaleInterface;
+use Pim\Component\Catalog\Model\ValueInterface;
+
+/**
+ * Simple factory of "incomplete values" collection.
+ *
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Completeness/IncompleteValueCollectionFactory.php]: `@internal`

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalizes missing values of a product model.
+ *
+ * @author    Julien Janvier (j.janvier@gmail.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ProductModelMissingValuesNormalizer implements NormalizerInterface
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php]: incomplete values

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+        $this->requiredValueCollectionFactory = $requiredValueCollectionFactory;
+        $this->incompleteValueCollectionFactory = $incompleteValueCollectionFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($productModel, $format = null, array $context = []): array
+    {
+        $family = $productModel->getFamily();
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductModelMissingValuesNormalizer.php]: useful?

File: src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php [hunk 18]
```diff
@@ -265,4 +110,22 @@ private function createCompleteness(
             $requiredCount
         );
     }
+
+    /**
+     * @param FamilyInterface $family
+     *
+     * @return Collection
+     */
+    private function getChannelsFromRequirements(FamilyInterface $family)
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php]: return type hint

File: src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollection.php [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @param ValueInterface[] $values
+     */
+    public function __construct(array $values)
+    {
+        $this->values = [];
+
+        foreach ($values as $value) {
+            $this->add($value);
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollection.php]: `InvalidArgumentException` versus `TypeException`

File: src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollectionFactory.php [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+use Pim\Component\Catalog\Completeness\Checker\ValueCompleteCheckerInterface;
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Pim\Component\Catalog\Model\EntityWithValuesInterface;
+use Pim\Component\Catalog\Model\LocaleInterface;
+use Pim\Component\Catalog\Model\ValueInterface;
+
+/**
+ * Simple factory of "incomplete values" collection.
+ *
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollectionFactory.php]: internal ?

File: src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollection.php [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * @param ValueInterface[] $values
+     */
+    public function __construct(array $values)
+    {
+        $this->values = [];
+
+        foreach ($values as $value) {
+            $this->add($value);
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollection.php]: idem for exception

File: src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollectionFactory.php [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+namespace Pim\Component\Catalog\EntityWithFamily;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Pim\Component\Catalog\Model\ChannelInterface;
+use Pim\Component\Catalog\Model\FamilyInterface;
+
+/**
+ * Simple factory of a "required value" collection.
+ *
```
Comment (aRn0D) [inline src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollectionFactory.php]: internal ?

File: src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php (hunk 18 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php]: done

File: src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollection.php (hunk 19 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollection.php]: done

File: src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollectionFactory.php (hunk 20 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/EntityWithFamily/IncompleteValueCollectionFactory.php]: done

File: src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollection.php (hunk 21 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollection.php]: done

File: src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollectionFactory.php (hunk 22 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/EntityWithFamily/RequiredValueCollectionFactory.php]: done

**Pull request merged**